x-pack/plugin/core/build.gradle

```diff
@@ -142,3 +142,9 @@ if (BuildParams.inFipsJvm){
   // Test clusters run with security disabled
   tasks.named("javaRestTest").configure{enabled = false }
 }
+
+if (BuildParams.isSnapshotBuild() == false) {
+  tasks.named("test").configure {
+    systemProperty 'es.untrusted_remote_cluster_feature_flag_registered', 'true'
```

Comment:
Need this for unit tests, e.g., in `x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptorTests.java`

Since only applies to unit tests and not broader integration or REST tests, I think it makes sense to enable the flag here instead of handling it on a more granular, case-by-case bases.

Reply:
Would it help to add `test-release` tag to this PR before you merge?

Reply:
Yup, it's there already.

Reply:
I don't think we want this here... The flag is already set for non-release (snapshot) builds, so it only impacts release builds this effectively says "for release testing, enable the feature flag" ... IMO, that is the exact opposite of what we want. We want snapshot testing to execute our tests as if there were no feature flag and the feature just exists, and we want release testing to execute our tests as if there was no feature at all. 

Some tests (especially unit) won't care about the feature flag and those should be executed, no harm no foul. But for the tests that do care about the feature flag they should not be executed as part of the release tests. They can be skipped with `assumeTrue` or `assumeFalse` per test and can be identified with the following additional command line arguments (or the `test-release` tag in GH).  

```
./gradlew :x-pack:plugin:core:test --tests "org.elasticsearch.xpack.core.security.authz.RoleDescriptorTests" -Dbuild.snapshot=false -Dtests.jvm.argline="-Dbuild.snapshot=false" -Dlicense.key=x-pack/plugin/core/src/test/resources/public.key
```

Reply:
Is it possible to set the flag and remove it before/after the specific test suite ? My main concern is that we are setting it for all x-pack core tests and we risk passing all of our x-pack tests but production code is actually broken. 

> The approach I'm using here is quite common in our codebase

I disagree with that approach unless they are trying to achieve something different. Why would you perform release testing and enable a feature that is not ready for release ? Maybe their feature is mostly ready for release and that is their intent ?

Reply:
> My main concern is that we are setting it for all x-pack core tests and we risk passing all of our x-pack tests but production code is actually broken.

Absolutely, I share that concern. My take on this is that we are lenient with unit tests to keep things simple and counting on bigger issues in release code to be caught by integration and rest tests (which I don't think are captured by the `test` task and therefore will not get the feature flag set).

That said, I agree that this doesn't seem like the right way to go. 

> Is it possible to set the flag and remove it before/after the specific test suite ? 

It should be possible. I see two ways here:

Tweak `UNTRUSTED_REMOTE_CLUSTER_FEATURE_FLAG_REGISTERED` such that it can be dynamically reset, and then add something like:

```
public void enableFeature() {
    AccessController.doPrivileged((PrivilegedAction<String>) () -> {
        if (TcpTransport.isUntrustedRemoteClusterEnabled()) {
            return null;
        }
        return System.setProperty("es.untrusted_remote_cluster_feature_flag_registered", "true");
    });
    resetFeatureFlag(); // Resets the feature flag by re-reading it from system prop
}
```

to a test suite and calling `enableFeature` for tests that need it. There would be extra quirks to work out here, such as setting the flag back to false if it was false before etc. Overall seems very brittle, esp. with a mutable static variable thrown into the mix.

The alternative would be to whitelist specific test suites inside build.gradle such that we only set the system prop inside `.configure` for specific test suites -- my gradle foo is decidedly not strong enough to say if that's a reasonable thing to do or if things like task configuration and test names shouldn't be conflated. I'll play around with this on a separate branch, because I'm curious to figure out how this would work.

Neither of these options are ideal, and both would establish a new pattern for how we work with feature flags which I would prefer to avoid diving into, at this stage.

I'm leaning towards refactoring the tests in this PR to explicitly account for the feature flag where needed and excluding tests with `assumeTrue` where sufficient. It adds some complexity but shouldn't be too bad. If you have time, would be great to sync around this on Slack/Zoom.

Reply:
I think we are bumping into some lack of support for features flags, unit tests, and release testing. That will probably take a while to unwind. As you mentioned bigger issues will more likely be caught in integration tests so I don't have a huge preference to address this now for unit tests. If there is a simple/fast/not-intrusive way to conditionally enable/disable , but if not that is OK.  We should revisit the support for feature flags in the future, but let this as-is for now.